File: src/main/java/shopping/product/service/ProductCommandService.java
package shopping.product.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import shopping.product.domain.Price;
import shopping.product.domain.Product;
import shopping.product.repository.ProductRepository;
import shopping.product.service.dto.ProductRegisterInput;

@Service
@Transactional
@RequiredArgsConstructor
public class ProductCommandService {
    private final ProductRepository productRepository;

    public Product register(ProductRegisterInput input) {
        return productRepository.save(Product.create(input.name(), input.price(), input.imageUrl()));
    }

    public Product update(Long id, ProductRegisterInput input) {
        Product product = getProduct(id);
        product.update(input.name(), new Price(input.price()), input.imageUrl());
        return product;
    }

    public void delete(Long id) {
        getProduct(id).delete();
    }

    private Product getProduct(Long id) {
        return productRepository.findByIdAndDeletedFalse(id)
                                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 상품입니다."));
    }
}

File: src/main/kotlin/shopping/product/application/ProductCommandService.kt
package shopping.product.application

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import shopping.global.exception.ApplicationException
import shopping.global.exception.ErrorCode
import shopping.global.infra.TransactionTemplate
import shopping.product.application.command.ProductCreateCommand
import shopping.product.application.command.ProductUpdateCommand
import shopping.product.domain.Product

@Service
class ProductCommandService(
    private val profanityValidator: ProfanityValidator,
    private val productCommandRepository: ProductCommandRepository,
    private val productQueryRepository: ProductQueryRepository,
    private val transactionTemplate: TransactionTemplate,
) {
    @Transactional
    fun createProduct(productCreateCommand: ProductCreateCommand): Product {
        validateProfanityProductName(productCreateCommand.productName)
        val product = productCreateCommand.toEntity()

        return productCommandRepository.save(product)
    }

    fun modifyProduct(productId: Long, productUpdateCommand: ProductUpdateCommand) {
        validateProfanityProductName(productUpdateCommand.productName)

        transactionTemplate.execute {
            val product = findProductNotDeleted(productId)

            product.modify(productUpdateCommand.toEntity())
        }
    }

    private fun validateProfanityProductName(productName: String) {
        if (profanityValidator.containsProfanity(productName)) {
            throw ApplicationException(ErrorCode.PROFANITY_CONTAIN_PRODUCT_NAME)
        }
    }

    @Transactional
    fun deletedProduct(productId: Long) {
        val product = findProductNotDeleted(productId)
        product.delete()
    }

    private fun findProductNotDeleted(productId: Long): Product {
        productQueryRepository.findByIdAndNotDeleted(productId)?.let { return it }
        throw ApplicationException(ErrorCode.PRODUCT_NOT_FOUND)
    }

}
